renderer/FrameBuffer.cpp

```diff
@@ -21,765 +21,311 @@ Project: The Dark Mod (http://www.thedarkmod.com/)
 }
 
-void FB_Clear() {
-	fboPrimary = fboResolve = fboPostProcess = fboShadowStencil = fboShadowAtlas = pbo = 0;
-	renderBufferColor.handle = renderBufferDepthStencil.handle = renderBufferPostProcess.handle = 0;
+void FrameBuffer::Generate() {
+	generator( this );
+	Validate();
 }
 
-void EnterPrimary() {
-	if ( r_softShadowsQuality.GetBool() ) {
-		r_useGLSL = true;
+void FrameBuffer::BindDraw() {
+	if (!initialized) {
+		Generate();
+		// definitely bind in this case since we recreated the fbo object
+		qglBindFramebuffer(GL_DRAW_FRAMEBUFFER, fbo);
 	}
-	depthCopiedThisView = false;
+
+	if (frameBuffers->activeDrawFbo != this) {
+		qglBindFramebuffer(GL_DRAW_FRAMEBUFFER, fbo);
+		frameBuffers->activeDrawFbo = this;
+	}
+}
 
-	if ( !r_useFbo.GetBool() ) {
-		return;
+void FrameBuffer::AddRenderBuffer( GLuint &buffer, GLenum attachment, GLenum format, const idStr &name ) {
+	if (buffer != 0) {
+		qglDeleteRenderbuffers(1, &buffer);
 	}
+	qglGenRenderbuffers(1, &buffer);
+	qglBindRenderbuffer(GL_RENDERBUFFER, buffer);
+	GL_SetDebugLabel( GL_RENDERBUFFER, buffer, name );
 
-	if ( primaryOn ) {
-		return;
+	if (msaa > 1) {
+		qglRenderbufferStorageMultisample(GL_RENDERBUFFER, msaa, format, width, height);
+	} else {
+		qglRenderbufferStorage(GL_RENDERBUFFER, format, width, height);
 	}
-	CheckCreatePrimary();
-
-	qglBindFramebuffer( GL_FRAMEBUFFER, fboPrimary );
-
-	primaryOn = true; // needs to set before GL_Viewport/GL_Scissor
-
-	GL_Viewport( tr.viewportOffset[0] + backEnd.viewDef->viewport.x1,	// FIXME: must not use tr in backend
-	             tr.viewportOffset[1] + backEnd.viewDef->viewport.y1,
-	             backEnd.viewDef->viewport.x2 + 1 - backEnd.viewDef->viewport.x1,
-	             backEnd.viewDef->viewport.y2 + 1 - backEnd.viewDef->viewport.y1 );
-
-	GL_Scissor( tr.viewportOffset[0] + backEnd.viewDef->viewport.x1 + backEnd.viewDef->scissor.x1,
-	            tr.viewportOffset[1] + backEnd.viewDef->viewport.y1 + backEnd.viewDef->scissor.y1,
-	            backEnd.viewDef->scissor.x2 + 1 - backEnd.viewDef->scissor.x1,
-	            backEnd.viewDef->scissor.y2 + 1 - backEnd.viewDef->scissor.y1 );
 
-	qglClear( GL_COLOR_BUFFER_BIT ); // otherwise transparent skybox blends with previous frame
-
-	GL_CheckErrors();
+	Bind();
+	qglFramebufferRenderbuffer(GL_FRAMEBUFFER, attachment, GL_RENDERBUFFER, buffer);
 }
 
-// switch from fbo to default framebuffer, copy content
-void LeavePrimary() {
-	if ( !primaryOn ) 
-		return;
-	primaryOn = false;
-	GL_CheckErrors();
-
-	GL_Viewport( 0, 0, glConfig.vidWidth, glConfig.vidHeight );
-	GL_Scissor( 0, 0, glConfig.vidWidth, glConfig.vidHeight );
-
-	if ( r_multiSamples.GetInteger() > 1 ) {
-		FB_ResolveMultisampling( GL_COLOR_BUFFER_BIT );
-		qglBindFramebuffer( GL_READ_FRAMEBUFFER, fboResolve );
+void FrameBuffer::AddRenderTexture( idImage *texture, GLenum attachment, int mipLevel ) {
+	if (texture->uploadWidth >> mipLevel != width || texture->uploadHeight >> mipLevel != height) {

```

Comment:
Bit operations without parentheses make me nervous =)

Reply:
Yeah, fair enough :) I'm too used to have a linter and auto-formatter to take care of these things :D